### ISSUE TITLE
Line wrapping broken (for me)

### ISSUE THREAD

**Issue by cvfosammmm** [xi-editor/xi-editor#1117]
Hi all,

I'm a fan of Xi and want to use it for my LaTeX editor (https://gitlab.gnome.org/cvfosammmm/setzer) and maybe other projects. I started work on a library that would make it easy to integrate Xi in Gtk based projects (it's written in c). I can't however get line wrapping to work. Here's two examples:

- Paste a couple of lines (code), say 300 lines. -> Xi approximates (I think) the line count and responds with a (wrong) number of invalid lines plus the lines on display containing the cursor. -> I scroll the cursor on screen. -> I update the scroll range (lines on display) and send to Xi. -> Xi calculates the real line numbers of a couple of logical lines and the document height changes (always seems to increase), sending an update where the cursor is often not included, scrolling it off screen.

- I show an area with invalid lines above. When I send the move_up message, those lines are wrapped and scrolled on screen in full (not just the last part).

There's other problems too.

So my question: how am I supposed to implement line wrapping correctly (best practice)?

PS: Would you consider adding an option to break line by word and then letter for long words? GtkScourceView has an option for that and I think it makes a lot of sense.

**Comment by Cogitri** [comment]
Hello there,

> I started work on a library that would make it easy to integrate Xi in Gtk based projects (it's written in c).

first of all, that seems like a very interesting project - please do ping me when you have something presentable, right now I have my own custom stuff for gxi, I'd be highly interested in having a library for interacting with Xi (I'd also like to help - but I barely know C, I only know Rust sort of well). I could lib-ize the parts of gxi that interact with xi directly (Rust FFI is great), I guess...

I'm not _quite_ so sure if I understand your problem correctly. If you have enabled word wrapping Xi should just send you a [measure_width](https://xi-editor.io/docs/frontend-protocol.html#measure_width) RPC to ask you how big a string will be when rendered. You have to send [resize](https://xi-editor.io/docs/frontend-protocol.html#resize) before that so that Xi knows how big your drawing (editing?) area is. That should be about what you have to do as frontend to get working word wrapping (or at least that did it for gxi):

![screenshot from 2019-02-16 22-04-59](https://user-images.githubusercontent.com/8766773/52905185-ee59dc00-3236-11e9-9a3a-47b5b151a56b.png)


 > PS: Would you consider adding an option to break line by word and then letter for long words? GtkScourceView has an option for that and I think it makes a lot of sense.

That does seem very useful, yes

**Comment by cvfosammmm** [comment]
Hi Cogitri,

the line wrapping itself isn't the problem. I reply to measure_width, send resize and scroll commands. But the scroll commands trigger recalculations of line count, which seems to be the main problem. That's why I get weird scrolling behaviour. For example if I stop sending scroll commands I can fix my first issue above, but obviously I can't do that, or else I wouldn't get updates when scrolling..

I think you could use my library with rust. It has gobject introspection bindings which should work in rust (I haven't tried). I'm myself using it with python. As to the current state: apart from line wrapping it works ok. Performance is very good. But many things are still missing. Before a release I definitely want line wrapping to work and do a bit of refactoring, fix couple things, ..

Unfortunately gxi doesn't really solve my problems. Sometimes it doesn't display lines after pasting (as in issue 1 above), sometimes it crashes. But it doesn't scroll the cursor off screen after pasting. I'm wondering how you do it. Could it be you're not sending the scroll command? Also sometimes in gxi the lines are not wrapped actually.

**Comment by Cogitri** [comment]
> Unfortunately gxi doesn't really solve my problems. Sometimes it doesn't display lines after pasting (as in issue 1 above), sometimes it crashes

Ah, yup, seems like my logical line handling is somewhat broken. Pushed a commit to at least fix the crashing.

> I think you could use my library with rust. It has gobject introspection bindings which should work in rust (I haven't tried).

Alright, will try.

**Comment by cvfosammmm** [comment]
Trying to explain my first problem a bit more:

me: paste code
xi: update - 679 invalid lines + 11 inserts
xi: scroll_to 689
me: scroll - 680:689
xi: update - 678 invalid + 1 insert + 679 skip + 1 copy +  11 inserts + 11 invalid lines

**Comment by cmyr** [comment]
Hmm, some quick observations:

xi bounds the number of lines it will expect the frontend to keep cached. This value is currently 1000; if a document has more than 1000 _visual lines_ (not logical lines; that is, a line that is wrapped once is a single logical line and two visual lines) than scrolling will cause some valid lines to be removed from the cache. I'm not sure this is your problem, but it might be worth knowing.

I'm confused by your last example, though. If you could come up with a minimal reproduction of your problem, I'd be happy to try and walk through it? And if you haven't already, it might be worth looking at the [line cache implementation in xi-mac](https://github.com/xi-editor/xi-mac/blob/master/Sources/XiEditor/LineCache.swift#L111).

**Comment by cvfosammmm** [comment]
Hi cmyr,

I don't think it's a problem with frontend side caching. The numbers above are a real observation and there were no other editing operations before in that session. Also even if the last lines were copied, the result would be the same. And even if there was an update of 1000 lines, what would happend if I paste 2000 lines? To reproduce just paste a long file with lots of line wrapping I guess.

Can you tell me if the physical lines in invalid areas are approximated in Xi? If so, I don't think my problem can be solved in the frontend easily (at all?), at least not without breaking asynchronicity.

I try to explain why: If I paste like above and the number of invalids at the start is not exact, how could I ever scroll to the top without some recalculation and after that a scrolling glitch where lines are pushed down by the newly "inserted" ones?

**Comment by cvfosammmm** [comment]
I think @raphlinus wrote somewhere that line wrapping should happen incrementally without the frontend waiting for upgrades and that an approximation is sent at first. Is that how it is currently implemented? I think this is generally a good way to do it, but I don't know how the frontend is supposed to handle it.

**Comment by cmyr** [comment]
I don't see this issue when pasting a big string with lots of line wrap into xi-mac.

The numbers are not approximate; core knows exactly how many lines are in the document.

Currently, how line wrapping works: after calculating what will be the visible region of the document, we immediately (before rendering) update that region's wrap state, and then in the background update other regions.

I wonder if this is your problem? There's currently a sort-of-bug where if you're in a very big document, and you resize the window, as the area above the viewport rewraps (asynchronously, a few ms later) it causes the viewport to move; you'll get sent scroll events, I believe. Could this be what's going on?

**Comment by cvfosammmm** [comment]
> .. and then in the background update other regions. I wonder if this is your problem?

Yes, I think that's my problem. The numbers have to be approximate in this case, no? If the document is not fully wrapped at the time of the first update.

I think that's not necessarily problematic if Xi sends me a message when additional lines have been wrapped. But I don't think it does. The xi-mac issue you describe seems related.

**Comment by cvfosammmm** [comment]
@cmyr Does the resize sort-of-bug have a github issue?

Can you confirm that my problem is a real bug in xi-editor? Or am I missing something? I don't understand swift at all so can't easily look for a fix in the xi-mac source. This is a blocker for me. If I can't get line wrapping to work my editor will not be very useful..

**Comment by cmyr** [comment]
@cvfosammmm I've opened #1135, which is basically a postmortem that describes the current problem.

**Comment by cvfosammmm** [comment]
@cmyr Thanks.
